File: ecommerce_ds/ecommerce_ex.py
# e_commerce.py
# Esqueleto do sistema de E-commerce em Python.
# Objetivo: fornecer somente as estruturas básicas + comentários orientando a implementação.
# Observações didáticas:
# - Para valores monetários, prefira 'decimal.Decimal' a 'float' (precisão).
# - IDs podem ser gerados com uuid.uuid4().
# - Se desejar, adicione validações (ex.: CPF, email) como extensão.

import uuid                     # gerar identificadores únicos (UUID)
from decimal import Decimal     # representação apropriada de dinheiro
from typing import List, Dict   # anotações de tipo (listas e dicionários)


# =========================
# CLASSE PESSOA
# =========================
class Pessoa:
    """
    Representa uma pessoa com atributos básicos.
    Atributos obrigatórios: nome, cpf, email.

    Tarefas do aluno:
    - Implementar o construtor (__init__) atribuindo os atributos.
    - (Opcional) Validar CPF / email.
    - (Opcional) Criar __str__/__repr__ para facilitar a depuração.
    """

    def __init__(self, nome: str, cpf: str, email: str) -> None:
        # TODO: atribuir os campos (ex.: self.nome = nome, etc.)
        # TODO (opcional): validar formato do cpf/email.
        pass


# =========================
# CLASSE CLIENTE
# =========================
class Cliente:
    """
    Representa um cliente do e-commerce, associado a uma Pessoa.
    Atributos: pessoa (objeto Pessoa), id_cliente (str).

    Tarefas do aluno:
    - Implementar o construtor recebendo uma Pessoa.
    - Gerar id_cliente com str(uuid.uuid4()).
    - (Opcional) Criar __str__/__repr__.
    """

    def __init__(self, pessoa: Pessoa) -> None:
        # TODO: self.pessoa = pessoa
        # TODO: self.id_cliente = str(uuid.uuid4())
        pass


# =========================
# CLASSE PRODUTO
# =========================
class Produto:
    """
    Representa um produto disponível para venda.
    Atributos: nome (str), preco (Decimal), estoque (int).

    Tarefas do aluno:
    - Implementar o construtor e armazenar os atributos.
    - (Opcional) Validar: preco >= 0 e estoque >= 0.
    - (Opcional) Criar __str__/__repr__.
    - (Opcional) Métodos utilitários (ex.: pode_atender, debitar, creditar) se o enunciado pedir.
    """

    def __init__(self, nome: str, preco: Decimal, estoque: int) -> None:
        # TODO: self.nome = nome
        # TODO: self.preco = preco (usar Decimal('199.90') ao instanciar)
        # TODO: self.estoque = estoque
        # TODO (opcional): validar não-negatividade de preco/estoque
        pass


# =========================
# CLASSE ITEMPEDIDO
# =========================
class ItemPedido:
    """
    Representa um item dentro de um pedido (produto + quantidade).
    Atributos: produto (Produto), quantidade (int).

    Tarefas do aluno:
    - Implementar o construtor.
    - Implementar calcular_total() = produto.preco * quantidade.
    """

    def __init__(self, produto: Produto, quantidade: int) -> None:
        # TODO: self.produto = produto
        # TODO: self.quantidade = quantidade
        # TODO (opcional): validar quantidade > 0
        pass

    def calcular_total(self) -> Decimal:
        """
        Retorna o subtotal do item:
        Fórmula: produto.preco * Decimal(quantidade).
        """
        # TODO: retornar self.produto.preco * Decimal(self.quantidade)
        raise NotImplementedError("Implementar cálculo do subtotal do item.")


# =========================
# CLASSE PEDIDO
# =========================
class Pedido:
    """
    Representa um pedido realizado por um cliente.
    Atributos: cliente (Cliente), itens (List[ItemPedido]).

    Tarefas do aluno:
    - Implementar o construtor inicializando a lista de itens vazia.
    - Implementar adicionar_item(produto, quantidade): cria ItemPedido e adiciona à lista.
    - Implementar calcular_total(): soma dos totais dos itens.
    - (Opcional) confirmar(): validar estoques e debitar (se fizer parte do enunciado).
    """

    def __init__(self, cliente: Cliente) -> None:
        # TODO: self.cliente = cliente
        # TODO: self.itens = []
        pass

    def adicionar_item(self, produto: Produto, quantidade: int) -> None:
        """
        Adiciona um ItemPedido à lista de itens.
        """
        # TODO: criar ItemPedido(produto, quantidade) e adicionar em self.itens
        raise NotImplementedError("Implementar inclusão de item no pedido.")

    def calcular_total(self) -> Decimal:
        """
        Calcula o total somando os subtotais dos itens.
        """
        # TODO: somar it.calcular_total() para cada item em self.itens
        raise NotImplementedError("Implementar cálculo do total do pedido.")


# =========================
# CLASSE MENU
# =========================
class Menu:
    """
    Gerencia o e-commerce: cadastro de clientes, produtos e pedidos.
    Sugestões de estruturas em memória:
    - clientes: Dict[str, Cliente]  (chave = CPF)
    - produtos: Dict[str, Produto]  (chave = ID ou nome; defina a estratégia)
    - pedidos: List[Pedido]

    Tarefas do aluno:
    - Implementar o construtor, inicializando as coleções.
    - Implementar métodos de cadastro e criação de pedidos.
    - Implementar exibir_menu() com interação via input() (CLI simples).
    """

    def __init__(self) -> None:
        # TODO: self.clientes: Dict[str, Cliente] = {}
        # TODO: self.produtos: Dict[str, Produto] = {}
        # TODO: self.pedidos: List[Pedido] = []
        pass

    def cadastrar_cliente(self, nome: str, cpf: str, email: str) -> None:
        """
        Cria Pessoa -> Cliente e armazena em self.clientes por CPF.
        """
        # TODO: implementar cadastro
        raise NotImplementedError("Implementar cadastro de cliente.")

    def cadastrar_produto(self, nome: str, preco: Decimal, estoque: int) -> None:
        """
        Cria Produto e armazena em self.produtos.
        Dica: se optar por chavear por ID, gere uma string com uuid.uuid4().
        """
        # TODO: implementar cadastro
        raise NotImplementedError("Implementar cadastro de produto.")

    def criar_pedido(self, cpf: str) -> None:
        """
        Fluxo sugerido:
        - Buscar cliente por CPF.
        - Permitir adicionar itens (loop de leitura).
        - Mostrar total.
        - (Opcional) Confirmar e debitar estoque (se previsto no enunciado).
        """
        # TODO: implementar criação do pedido conforme a proposta da disciplina
        raise NotImplementedError("Implementar criação de pedido.")

    def listar_pedidos(self) -> None:
        """
        Exibe um resumo dos pedidos armazenados.
        """
        # TODO: iterar por self.pedidos e imprimir/formatar
        raise NotImplementedError("Implementar listagem de pedidos.")

    def exibir_menu(self) -> None:
        """
        Exibe o menu e executa as opções escolhidas (CLI).
        Dica: usar input() e um loop while para o menu principal.
        """
        # TODO: implementar interface de linha de comando
        raise NotImplementedError("Implementar menu interativo.")


# =========================
# PONTO DE ENTRADA
# =========================
if __name__ == "__main__":
    # Instancia o Menu e inicia a CLI.
    # Aluno deve implementar os métodos do Menu para que a interação funcione.
    menu = Menu()
    menu.exibir_menu()

File: ecommerce_ds/Ecommerce_ex.java
// Main.java
// Esqueleto do sistema de E-commerce em Java, convertido do template Python.
// Objetivo: fornecer somente as estruturas básicas + comentários orientando a implementação.
// Obs.: para valores monetários, prefira BigDecimal a double.

import java.math.BigDecimal;     // recomendado para preços (evita erro de ponto flutuante)
import java.util.*;              // List, Map, UUID, etc. (use conforme necessário)

/* =========================
   CLASSE PESSOA
   ========================= */
/**
 * Representa uma pessoa com atributos básicos.
 * Atributos: nome, cpf, email.
 * Tarefas do aluno:
 *  - Implementar construtor
 *  - (Opcional) Validar CPF / email
 *  - Criar getters/setters conforme necessidade
 */
class Pessoa {
    // Atributos (privados para encapsulamento)
    private String nome;   // nome completo da pessoa
    private String cpf;    // documento (string)
    private String email;  // email de contato

    // Construtor: receber nome, cpf, email e armazenar nos atributos
    public Pessoa(String nome, String cpf, String email) {
        // TODO: atribuir campos e (opcional) validar entradas
    }

    // Getters/Setters (implementar conforme necessidade didática)
    // public String getNome() { ... }
    // public void setNome(String nome) { ... }
    // public String getCpf() { ... }
    // public void setCpf(String cpf) { ... }
    // public String getEmail() { ... }
    // public void setEmail(String email) { ... }

    // (Opcional) toString para facilitar depuração/listagem
    // @Override public String toString() { ... }
}

/* =========================
   CLASSE CLIENTE
   ========================= */
/**
 * Representa um cliente do e-commerce, associado a uma Pessoa.
 * Atributos: pessoa (composição), idCliente (String).
 * Tarefas do aluno:
 *  - Implementar construtor recebendo Pessoa
 *  - Gerar idCliente via UUID.randomUUID().toString()
 *  - Criar getters conforme necessidade
 */
class Cliente {
    private Pessoa pessoa;       // dados da pessoa
    private String idCliente;    // identificador único (gerado por UUID)

    public Cliente(Pessoa pessoa) {
        // TODO: armazenar pessoa
        // TODO: this.idCliente = UUID.randomUUID().toString();
    }

    // Getters necessários
    // public Pessoa getPessoa() { ... }
    // public String getIdCliente() { ... }

    // @Override public String toString() { ... }
}

/* =========================
   CLASSE PRODUTO
   ========================= */
/**
 * Representa um produto disponível para venda.
 * Atributos: nome, preco (BigDecimal), estoque (int).
 * Tarefas do aluno:
 *  - Implementar construtor
 *  - (Opcional) Validar preço >= 0 e estoque >= 0
 *  - Getters/Setters necessários (ex.: atualizar preço)
 */
class Produto {
    private String nome;           // nome do produto
    private BigDecimal preco;      // preço unitário (usar BigDecimal)
    private int estoque;           // quantidade disponível

    public Produto(String nome, BigDecimal preco, int estoque) {
        // TODO: atribuir campos e (opcional) validar valores
    }

    // Getters/Setters conforme necessidade
    // public String getNome() { ... }
    // public void setNome(String nome) { ... }
    // public BigDecimal getPreco() { ... }
    // public void setPreco(BigDecimal preco) { ... }
    // public int getEstoque() { ... }
    // public void setEstoque(int estoque) { ... }

    // (Opcional) métodos utilitários como podeAtender(qtd), debitar/creditar
    // mas deixe para o momento da confirmação do pedido, se o enunciado exigir.
}

/* =========================
   CLASSE ITEMPEDIDO
   ========================= */
/**
 * Representa um item dentro de um pedido.
 * Contém um produto e uma quantidade.
 * Atributos: produto, quantidade.
 * Tarefas do aluno:
 *  - Implementar construtor
 *  - Implementar calcularTotal(): preco * quantidade
 */
class ItemPedido {
    private Produto produto;   // referência ao produto escolhido
    private int quantidade;    // quantidade solicitada

    public ItemPedido(Produto produto, int quantidade) {
        // TODO: atribuir campos e (opcional) validar quantidade > 0
    }

    /**
     * Retorna o subtotal deste item.
     * Fórmula: produto.preco * quantidade.
     */
    public BigDecimal calcularTotal() {
        // TODO: retornar produto.getPreco().multiply(BigDecimal.valueOf(quantidade));
        return null; // placeholder
    }

    // Getters necessários
    // public Produto getProduto() { ... }
    // public int getQuantidade() { ... }
}

/* =========================
   CLASSE PEDIDO
   ========================= */
/**
 * Representa um pedido realizado por um cliente.
 * Contém o cliente e uma lista de itens.
 * Atributos: cliente, itens (List<ItemPedido>).
 * Tarefas do aluno:
 *  - Implementar construtor (lista vazia inicialmente)
 *  - Implementar adicionarItem(produto, quantidade)
 *  - Implementar calcularTotal(): soma dos totais dos itens
 *  - (Opcional) confirmar(): validar estoque e debitar (se fizer parte do enunciado)
 */
class Pedido {
    private Cliente cliente;            // quem realizou o pedido
    private List<ItemPedido> itens;     // itens que compõem o pedido

    public Pedido(Cliente cliente) {
        // TODO: atribuir cliente e inicializar itens = new ArrayList<>();
    }

    /**
     * Adiciona um item ao pedido a partir de um produto e uma quantidade.
     */
    public void adicionarItem(Produto produto, int quantidade) {
        // TODO: criar ItemPedido e adicionar à lista
    }

    /**
     * Calcula o total do pedido somando os subtotais dos itens.
     */
    public BigDecimal calcularTotal() {
        // TODO: iterar pelos itens e somar it.calcularTotal()
        return null; // placeholder
    }

    // Getters necessários
    // public Cliente getCliente() { ... }
    // public List<ItemPedido> getItens() { ... }
}

/* =========================
   CLASSE MENU
   ========================= */
/**
 * Gerencia o e-commerce, permitindo o cadastro de clientes, produtos e pedidos.
 * Atributos: clientes, produtos, pedidos (coleções em memória).
 * Dicas:
 *  - Para busca rápida de cliente por CPF, use Map<String, Cliente>.
 *  - Para produtos, pode usar Map<String, Produto> (por ID) ou armazenar uma lista e buscar por nome.
 *  - Para pedidos, use List<Pedido>.
 * Tarefas do aluno:
 *  - Implementar os métodos conforme descrito no template Python original
 *  - Definir uma CLI simples em exibirMenu() (Scanner + switch)
 */
class Menu {
    // Estruturas de dados sugeridas
    private Map<String, Cliente> clientes;  // por CPF
    private Map<String, Produto> produtos;  // por ID ou por nome (defina a chave)
    private List<Pedido> pedidos;           // lista de pedidos

    public Menu() {
        // TODO: inicializar as coleções (ex.: new HashMap<>(), new ArrayList<>())
    }

    /**
     * Cadastra um cliente a partir de nome, cpf e email.
     */
    public void cadastrarCliente(String nome, String cpf, String email) {
        // TODO: criar Pessoa, criar Cliente e inserir em 'clientes' por CPF
    }

    /**
     * Cadastra um produto a partir de nome, preço e estoque.
     */
    public void cadastrarProduto(String nome, BigDecimal preco, int estoque) {
        // TODO: criar Produto e inserir em 'produtos'
        // Dica: se usar Map por ID, gere um ID (UUID) dentro do Produto
    }

    /**
     * Cria um pedido para o CPF informado.
     * Fluxo sugerido:
     *  - Buscar Cliente por CPF
     *  - Permitir adicionar itens (loop de leitura)
     *  - Mostrar total
     *  - (Opcional) Confirmar e debitar estoque
     */
    public void criarPedido(String cpf) {
        // TODO: implementar interação conforme proposta da disciplina
    }

    /**
     * Lista pedidos (resumo).
     */
    public void listarPedidos() {
        // TODO: iterar sobre 'pedidos' e imprimir um resumo
    }

    /**
     * Exibe menu e executa as opções escolhidas (CLI).
     * Dica: usar Scanner para ler opções e dados.
     */
    public void exibirMenu() {
        // TODO: laço principal (while), imprimir opções e chamar métodos acima
    }
}

/* =========================
   PONTO DE ENTRADA
   ========================= */
/**
 * Método main: cria o Menu e chama exibirMenu().
 * Aluno deve implementar a lógica interna do Menu.
 */
public class Ecommerce_ex {
    public static void main(String[] args) {
        Menu menu = new Menu();   // TODO: certifique-se de inicializar as coleções no construtor
        menu.exibirMenu();        // TODO: implementar fluxo do menu
    }
}

